File: apps/api/internal/domain/organization/organization.go
package organization

import (
	"context"
	"encoding/json"
	"time"
)

type Organization struct {
	ID        string          `db:"id"         json:"id"`
	Name      string          `db:"name"       json:"name"`
	Slug      string          `db:"slug"       json:"slug"`
	LogoURL   *string         `db:"logo_url"   json:"logo_url"`
	PlanTier  string          `db:"plan_tier"  json:"plan_tier"`
	Settings  json.RawMessage `db:"settings"   json:"settings"`
	CreatedAt time.Time       `db:"created_at" json:"created_at"`
	UpdatedAt time.Time       `db:"updated_at" json:"updated_at"`

	// Billing state (1:1 with the organization).
	PlanRenewsAt       *time.Time `db:"plan_renews_at"       json:"plan_renews_at"`
	BillingStatus      string     `db:"billing_status"       json:"billing_status"`
	BillingProvider    string     `db:"billing_provider"     json:"billing_provider"`
	BillingCustomerRef *string    `db:"billing_customer_ref" json:"billing_customer_ref"`
	PaymentMethodBrand *string    `db:"payment_method_brand" json:"payment_method_brand"`
	PaymentMethodLast4 *string    `db:"payment_method_last4" json:"payment_method_last4"`
}

type CreateInput struct {
	Name    string
	Slug    string
	OwnerID string
}

type UpdateInput struct {
	Name     *string
	LogoURL  *string
	Settings json.RawMessage
}

// PlanUpdate carries the billing fields applied when a plan change completes.
type PlanUpdate struct {
	PlanTier  string
	RenewsAt  *time.Time
}

// Usage is a point-in-time snapshot of an organization's resource consumption.
type Usage struct {
	Members          int64 `db:"members"            json:"members"`
	TicketsThisMonth int64 `db:"tickets_this_month" json:"tickets_this_month"`
	StorageBytes     int64 `db:"storage_bytes"      json:"storage_bytes"`
	Teams            int64 `db:"teams"              json:"teams"`
	Categories       int64 `db:"categories"         json:"categories"`
	APIKeys          int64 `db:"api_keys"           json:"api_keys"`
}

type Repository interface {
	Create(ctx context.Context, org *Organization) error
	FindByID(ctx context.Context, id string) (*Organization, error)
	FindBySlug(ctx context.Context, slug string) (*Organization, error)
	Update(ctx context.Context, id string, input UpdateInput) (*Organization, error)
	UpdatePlan(ctx context.Context, id string, input PlanUpdate) (*Organization, error)
	SlugExists(ctx context.Context, slug string) (bool, error)
}

// UsageRepository aggregates an organization's current resource usage.
type UsageRepository interface {
	Snapshot(ctx context.Context, orgID string) (*Usage, error)
}

File: apps/web/src/lib/api/organization.ts
import { api } from './client';

export type PlanTier = 'free' | 'pro' | 'business' | 'enterprise';

export interface PlanLimits {
	seats: number;
	tickets_per_month: number;
	storage_bytes: number;
	teams: number;
	categories: number;
	api_keys: number;
}

export interface Plan {
	tier: PlanTier;
	name: string;
	price_cents: number;
	currency: string;
	limits: PlanLimits;
}

export interface Usage {
	members: number;
	tickets_this_month: number;
	storage_bytes: number;
	teams: number;
	categories: number;
	api_keys: number;
}

export interface Organization {
	id: string;
	name: string;
	slug: string;
	logo_url: string | null;
	plan_tier: PlanTier;
	created_at: string;
	updated_at: string;
	plan_renews_at: string | null;
	billing_status: string;
	billing_provider: string;
	billing_customer_ref: string | null;
	payment_method_brand: string | null;
	payment_method_last4: string | null;
}

export type SessionStatus = 'pending' | 'completed' | 'cancelled' | 'failed' | 'expired';

export interface PaymentSession {
	id: string;
	org_id: string;
	from_tier: string | null;
	to_tier: PlanTier;
	status: SessionStatus;
	amount_cents: number;
	proration_cents: number;
	currency: string;
	provider: string;
	provider_ref: string | null;
	created_by: string | null;
	created_at: string;
	completed_at: string | null;
	expires_at: string | null;
}

export interface OrgOverview {
	organization: Organization;
	plan: Plan;
	usage: Usage;
	pending_session: PaymentSession | null;
}

export const organizationApi = {
	getOverview: () => api.get<OrgOverview>('/api/v1/organization'),

	listPlans: () => api.get<Plan[]>('/api/v1/organization/plans'),

	update: (input: { name: string }) => api.patch<Organization>('/api/v1/organization', input),

	initiatePlanChange: (toTier: PlanTier) =>
		api.post<PaymentSession>('/api/v1/organization/plan/sessions', { to_tier: toTier }),

	confirmSession: (id: string) =>
		api.post<PaymentSession>(`/api/v1/organization/plan/sessions/${id}/confirm`, {}),

	cancelSession: (id: string) =>
		api.post<void>(`/api/v1/organization/plan/sessions/${id}/cancel`, {}),

	sessionHistory: () => api.get<PaymentSession[]>('/api/v1/organization/plan/sessions')
};

const UNLIMITED = -1;

export function isUnlimited(value: number): boolean {
	return value < 0;
}

export function formatBytes(bytes: number): string {
	if (bytes < 0) return '∞';
	if (bytes === 0) return '0 B';
	const units = ['B', 'KB', 'MB', 'GB', 'TB'];
	const i = Math.floor(Math.log(bytes) / Math.log(1024));
	const value = bytes / Math.pow(1024, i);
	return `${value.toFixed(value >= 10 || i === 0 ? 0 : 1)} ${units[i]}`;
}

export function formatLimit(value: number, isBytes = false): string {
	if (isUnlimited(value)) return '∞';
	return isBytes ? formatBytes(value) : value.toLocaleString();
}

export function formatPrice(cents: number, currency = 'BRL'): string {
	return new Intl.NumberFormat('pt-BR', { style: 'currency', currency }).format(cents / 100);
}

export function usagePercent(current: number, limit: number): number {
	if (isUnlimited(limit) || limit === 0) return 0;
	return Math.min(100, Math.round((current / limit) * 100));
}

export { UNLIMITED };
